### PR TITLE
[ssr] extended intro patterns (episode 2)

### PR DESCRIPTION
**Pull request by gares**
Depends on: #6705 

This PR enables `=> ( | )` and `=> /2/` and `/2=` and has been extracted out of #6705 as per https://github.com/coq/coq/pull/6705#issuecomment-446229018

### CODE REVIEW

**Comment by gares** [comment]
Comments from the math-comp WG (@amahboubi @ggonthier @maximedenes)
- `move=> ... // ()` to test linearity can be written `move=> ... // => [-]` without the new dispatch ipat
- missing use case for `move=> ... (..|..|..)`
- numerical id in `/2/` or `/2=` is not very intelligible

IIRC the dispatch ipat was proposed by @charguer that is welcome to provide use cases.
We don't lack use cases for `/2=` and `/2/` but we lack a better syntax, suggestions are welcome.

File: doc/sphinx/proof-engine/ssreflect-proof-language.rst [hunk 1]
```diff
@@ -1564,10 +1564,10 @@ whose general syntax is
    i_view ::= {? %{%} } /@term %| /ltac:( @tactic )
 
 .. prodn::
-   i_pattern ::= @ident %| > @ident %| _ %| ? %| * %| + %| {? @occ_switch } -> %| {? @occ_switch }<- %| [ {?| @i_item } ] %| - %| [: {+ @ident } ]
+   i_pattern ::= @ident %| > @ident %| _ %| ? %| * %| + %| {? @occ_switch } -> %| {? @occ_switch }<- %| [ {?| @i_item } ] %| ( {?| @i_item } ) %| - %| [: {+ @ident } ]
 
 .. prodn::
-   i_block ::= [^ @ident ] %| [^~ @ident ]
+   i_block ::= [^ @ident ] %| [^~ @ident ] %| [^~ @num ] %|(^ @ident ) %| (^~ @ident ) %| (^~ @num )
```
Comment (gares) [inline doc/sphinx/proof-engine/ssreflect-proof-language.rst]: The last `[^ ... ]` should go in the other PR

File: doc/sphinx/proof-engine/ssreflect-proof-language.rst [hunk 2]
```diff
@@ -1787,6 +1802,13 @@ are all equivalent.
 
 :n:`[^~ @ident ]`
   *block destructing* using :token:`ident` as a suffix.
+:n:`[^~ @num ]`
```
Comment (gares) [inline doc/sphinx/proof-engine/ssreflect-proof-language.rst]: idem

File: doc/sphinx/proof-engine/ssreflect-proof-language.rst [hunk 3]
```diff
@@ -5369,7 +5395,7 @@ view :ref:`introduction_ssr`
 intro block :ref:`introduction_ssr`
 
 .. prodn::
-   i_block ::= [^ @ident ] %| [^~ @ident ]
+   i_block ::= [^ @ident ] %| [^~ @ident ] %| [^~ @num ] %|(^ @ident ) %| (^~ @ident ) %| (^~ @num )
```
Comment (gares) [inline doc/sphinx/proof-engine/ssreflect-proof-language.rst]: idem

**Comment by maximedenes** [comment]
@gares what's the status of this? Does it simply need a review and a merge?

**Comment by maximedenes** [comment]
In particular, were the remarks from the math comp meeting addressed?

**Comment by gares** [comment]
This needs more work (design work)

**Comment by maximedenes** [comment]
> This needs more work (design work)

Ok, thanks!

**Comment by Zimmi48** [comment]
Dear PR author(s) and contributors, it seems to us this PR may need additional work beyond the 8.12 branching deadline on May 15th.

We are thus removing the 8.12+beta1 milestone; please retarget as appropriate, and also consider updating labels to reflect current status, moving the PR to draft status, etc.

We apologize in advance if we misunderstood the PR status.

**Comment by charguer** [comment]
> IIRC the dispatch ipat was proposed by @charguer that is welcome to provide use cases.

It's been a while. From what I recall, one use case it related to the use of tactics as views, e.g. `/ltac:mytactic`. Such a tactic can produce one or more side conditions, and you may want to eliminate these sideconditions on the fly before going on with the chain of intropatterns. 

Another use case I think is when you do a on-the-fly destruct of a list (or a nat), and you want to invoke a tactic for the base case and continue with the intropatterns to handle the inductive case. 

What could/should be documented in that case is whether  `/ltac:(foo; [bar1|bar2])` is equivalent to `/ltac:foo (bar1|bar2)`. And whether  `/ltac:(foo1; [foo2; [bar1|bar2] | foo2; [bar3|bar4] ] )`  can be equivalently written  `/ltac:foo1 /ltac:foo2: (bar1|bar2|bar3|bar4)`.

**Pull request closed without merging**